### PR TITLE
Ensemble algorithm and code update for corrected data release 

### PR DESCRIPTION
**Pull request by lmcarboneau**
**Please don't merge this yet! It needs a code review, especially for the files outside of the ensemble!**

There are changes to the code that I'm concerned might conflict across methods, including:

1. updating the `taskmanager` to accept a user specified data source (to force FFI preference)

2. updating `run_tesscorr` to accept a user specified data source; updating to also accept a user specified `output_folder,` and to pass a user specified debug flag to the `BaseCorrector`

3. updating the `BaseCorrector` to accept the new parameters above, and implementing internal changes to use them properly (init: `self.output_folder = output_folder`;  `save_lightcurve()` and plots now save to `self.output_folder` when specified, etc)

4. updating the BaseCorrector to use `lc_corr` instead of `lc` (original object) :: there was some bug that occasionally didn't make a new object when it wasn't implicit, but it also increases the readability of the code in this section

5. passing `Tmag` along to the ensemble corrector :: I haven't checked if quality flags are also being passed (since they weren't available in the data I had earlier) but the ensemble corrector will eventually use both, so this is something we're making use of and not just a placeholder

6. commenting out two places where the code crashed (on our current data) :: first, the `rms_timescale()` function was giving me a type error; for some reason the `Binned_statistic` isn't getting an integer array for its bins? I'm looking into why this is popping up, but if there's already a fix for this we should get it in before the merge
the second is just that we don't have quality or error estimations to save to the FITS file (yet)

### CODE REVIEW

**Comment by lmcarboneau** [comment]
TravisCI is probably going to fail because the tests need to be updated. 
Current TO-DO list looks something like:

- [ ] update tests to include the `BaseCorrector` changes, pending the code review comments
- [ ] update the ensemble algorithm one more time, which will include error estimates
- [ ] determine whether or not quality flags will be included in this release version (update if needed)
- [ ] fix whatever weirdness is going on with the rms_hour() calculation (I think something is going sideways in scipy, but I haven't found the root of the issue yet)

**Comment by rhandberg** [comment]
Hi @lmcarboneau ,

Good work! Sorry it has taken me this long to respond...

I can see that most of the changes you have made to the "not ensemble" part of the code, we had actually already implemented in the CBV branch... That is now merged into the "devel" branch, so may I suggest that you discard the changes you made to the "not ensemble" parts and merge in the newest version of the devel branch... Then those things will no longer be in conflict...

After that, there are only a few small things, that needs renaming, but we can take those things after the above mentioned step...

File: cbv_spiketest.py [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+from scipy.interpolate import interp1d
+import astropy.units as u
+from scipy.interpolate import pchip_interpolate
+
+from corrections.cbv_corrector.cbv_main import CBV
+import corrections
+
+cbv_area = 334
+#data_folder= '/home/mikkelnl/ownCloud/Documents/Asteroseis/TESS/TASOC_code/corrections/corrections/data/cbv_old'
+data_folder= '/home/mikkelnl/ownCloud/Documents/Asteroseis/TESS/TASOC_code/corrections/corrections/data/cbv'
```
Comment (ojhall94) [inline cbv_spiketest.py:25]: Is this code (`cbv_spiketest.py`) just a test on a local machine for @Miklnl ? If it is we should probably flag it as such in the docstring at the top of the script.

Comment (rhandberg) [reply cbv_spiketest.py:25]: Yes... But it is fixed in the newest version of the devel branch.... The issue is really that this PR should not change anything not having something to do with the ensemble code... So any changes to the CBV code, `BaseCorrector`, `TaskManager` or `run_tesscorr` should properly be reverted

File: corrections/BaseCorrector.py [hunk 2]
```diff
@@ -394,23 +424,41 @@ def save_lightcurve(self, lc, output_folder=None):
...
+			os.chmod(save_file, 0o640) 
 
 			# Open the FITS file to overwrite the corrected flux columns:
-			with fits.open(save_file, mode='update') as hdu:
+			with fits.open(os.path.join(self.input_folder,fname), mode='update') as hdu:
 				# Overwrite the corrected flux columns:
 				hdu['LIGHTCURVE'].data['FLUX_CORR'] = lc.flux
-				hdu['LIGHTCURVE'].data['FLUX_CORR_ERR'] = lc.flux_err
-				hdu['LIGHTCURVE'].data['QUALITY'] = lc.quality
+				#hdu['LIGHTCURVE'].data['FLUX_CORR_ERR'] = lc.flux_err
```
Comment (ojhall94) [inline corrections/BaseCorrector.py:460]: Are these supposed to remain commented out in the final version?

File: corrections/BaseCorrector.py [hunk 3]
```diff
@@ -422,7 +470,7 @@ def save_lightcurve(self, lc, output_folder=None):
 						hdu['LIGHTCURVE'].header[key] = (value, lc.meta['additional_headers'].comments[key])
 
 				# Save the updated FITS file:
-				hdu.flush()
+				# hdu.flush()
```
Comment (ojhall94) [inline corrections/BaseCorrector.py:473]: Is this supposed to remain commented out in the final version?

File: corrections/BaseCorrector.py (hunk 2 above)
Comment (rhandberg) [reply corrections/BaseCorrector.py:460]: No... Again, an example of a change in this PR that we should revert

File: corrections/BaseCorrector.py [hunk 4]
```diff
@@ -457,4 +505,4 @@ def save_lightcurve(self, lc, output_folder=None):
 		# Store the output file in the details object for future reference:
 		save_file = os.path.relpath(save_file, output_folder).replace('\\', '/')
 
-		return save_file
+		return save_file
```
Comment (ojhall94) [inline corrections/BaseCorrector.py:508]: Is there a little weird symbol for everyone else too?

File: corrections/ensemble.py [hunk 5]
```diff
@@ -47,29 +70,43 @@ def do_correction(self, lc):
...
+        full_start = time.time()
 
-        # Calculate extra data for the target lightcurve
+        # Clean up the lightcurve by removing nans and ignoring data points with bad quality flags
+        og_time = lc.time.copy()
         lc = lc.remove_nans()
-        # NOTE: Is frange supposed to be missing the brackets and not join the percentiles ??
-        frange = np.percentile(lc.flux, 95) - np.percentile(lc.flux, 5) / np.mean(lc.flux)
+        lc_quality_mask = (lc.quality == 0)
+        # lc.time = lc.time[lc_quality_mask]
```
Comment (ojhall94) [inline corrections/ensemble.py:82]: Should these remain commented out in the final version?

File: corrections/ensemble.py [hunk 6]
```diff
@@ -97,208 +135,237 @@ def do_correction(self, lc):
...
+                    return None, STATUS.SKIPPED
                 search_loop.append("todolist.starid={:}".format(starid[next_star_index]))
                 next_star_task = self.search_database(search=search_loop, select=select_loop)[0]
                 next_star_lc = self.load_lightcurve(next_star_task).remove_nans()
                 search_loop.pop(-1)
                 
-                # Compute the rest of its data. NOTE: Change this to the database or not ???
-                frange = np.percentile(next_star_lc.flux, 95) - np.percentile(next_star_lc.flux, 5) / np.mean(next_star_lc.flux)
+                next_star_lc_quality_mask = (next_star_lc.quality == 0)
+                # next_star_lc.time = next_star_lc.time[next_star_lc_quality_mask]
```
Comment (ojhall94) [inline corrections/ensemble.py:164]: Should these remain commented out in the final version?

File: cbv_spiketest.py (hunk 1 above)
Comment (ojhall94) [reply cbv_spiketest.py:25]: Got it!

File: .travis.yml [hunk 7]
```diff
@@ -1,6 +1,5 @@
 language: python
 python:
```
Comment (ojhall94) [inline .travis.yml:2]: `.travis.yml` should be checked out from the PR.

File: cbv_spiketest.py [hunk 8]
```diff
@@ -0,0 +1,181 @@
+#!/usr/bin/env python3
```
Comment (ojhall94) [inline cbv_spiketest.py:1]: `cbv_spiketest.py` should be checked out from the PR

File: corrections/__init__.py [hunk 9]
```diff
@@ -4,6 +4,7 @@
 from __future__ import absolute_import
```
Comment (ojhall94) [inline corrections/__init__.py:4]: `__init__.py` should be checked out from the PR

File: corrections/cbv_corrector/CBVCorrector.py [hunk 10]
```diff
@@ -0,0 +1,775 @@
+#!/usr/bin/env python
```
Comment (ojhall94) [inline corrections/cbv_corrector/CBVCorrector.py:1]: `CBVCorrector.py`  should be checked out from the PR

File: corrections/cbv_corrector/GOC_code.py [hunk 11]
```diff
@@ -0,0 +1,456 @@
+#!/usr/bin/env python3
```
Comment (ojhall94) [inline corrections/cbv_corrector/GOC_code.py:1]: `GOC_code.py`  should be checked out from the PR

**Comment by ojhall94** [comment]
Right I have more of a grasp on the situation now. Something seems to have gone wrong during the rebase: all files apart from the EnsembleCorrector content should be up to date with the latest in `devel`, but all files are still being flagged as having changes with respect to `devel`.

I think the right course of action would be a merge `devel` -> `ensemble_comments` and in cases of merge conflicts prefer the `devel` version in all cases except for the Ensemble code content.

I'll have a swing at this tomorrow morning.

**Comment by ojhall94** [comment]
Okay, something funky is happening. The only differences between the current state of `devel` and `ensemble_comments` are:

`corrections/ensemble.py`
`corrections/BaseCorrector.py`
`run_tesscor.py`

Yet it insists on flagging 24 files as changed with respect to devel.
I've done my best to remedy whats going on but its gone beyond my Git knowledge. I think its time to abandon branch (@lmcarboneau) for a new PR off the devel branch and just refer back to this thread as the commit history on the ensemble work.

**Comment by rhandberg** [comment]
I think I second that @ojhall94... It seems that would be the easiest way out, as this also goes way beyond my Git-knowledge! So a new PR with just the changes to `ensemble.py`, based off the current devel branch.

**Comment by rhandberg** [comment]
This PR has been bypassed by https://github.com/tasoc/corrections/pull/14 and is therefore closed.

**Pull request closed without merging**